### ISSUE TITLE
[BUG] `TypeError` when using type annotations in signatures with defaults

### ISSUE THREAD

**Issue by P403n1x87** [cython/cython#6069]
### Describe the bug

When using type annitation on function signatures, the generated type checks are strict (`*_CheckExact`) and this leads to `TypeError` when passing a subclass of the specified type

### Code to reproduce the behaviour:

```python
class _UNSET(int):
    pass

UNSET = _UNSET()

class Foo:
    def __init__(
        self,
        name: str,
        arg: int = UNSET,
    )
```
yields
```
TypeError: Expected int, got _UNSET
```
This is the C check that gets generated
```c
if (!(likely(__Pyx_Py3Int_CheckExact(__pyx_t_8))||((__pyx_t_8) == Py_None) || __Pyx_RaiseUnexpectedTypeError("int", __pyx_t_8))) __PYX_ERR(0, 87, __pyx_L1_error)
```

### Expected behaviour

No exception

### OS

macOS

### Python version

3.11.4

### Cython version

3.0.9

**Comment by scoder** [comment]
Why do you think this is a bug?

**Comment by da-woods** [comment]
Duplicate of https://github.com/cython/cython/issues/5908 and https://github.com/cython/cython/issues/5853 (and maybe others).

This behaviour is intended and documented in a couple of places https://cython.readthedocs.io/en/latest/src/tutorial/pure.html#reference-table and https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html#annotation-typing.

There's a better explanatory error message (https://github.com/cython/cython/pull/5915), but not in the 3.0 releases unfortunately.

Essentially Cython uses the annotations to generate faster code, and this means it has to be stricter than PEP-484 with exact types. If this isn't what you want then you should turn the annotation typing directive off.

**Comment by P403n1x87** [comment]
@da-woods thanks for the links. My goal was to take a pure Python codebase and Cythonise it without modifying the code. Is there a chance that annotation typing may be disabled globally, say in `setup.py`?

**Comment by da-woods** [comment]
Yes you can set compiler directives in setup.py https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#in-setup-py. The directive you want is `annotation_typing`
